src/sota_tools/deploy.cc

```diff
@@ -143,37 +144,40 @@ bool copy_repo(const std::string &cacerts, const std::string &src, const std::st
 
   LOG_INFO << "Uploaded " << uploaded << " objects";
   LOG_INFO << "Already present " << present_already << " objects";
-  // Push ref
 
-  if (root_object->is_on_server() == OBJECT_PRESENT) {
-    CURL *easy_handle = curl_easy_init();
-    curl_easy_setopt(easy_handle, CURLOPT_VERBOSE, get_curlopt_verbose());
-    ostree_ref.PushRef(push_server, easy_handle);
-    CURLcode err = curl_easy_perform(easy_handle);
-    if (err) {
-      LOG_ERROR << "Error pushing root ref:" << curl_easy_strerror(err);
-      errors++;
-    }
-    long rescode;
-    curl_easy_getinfo(easy_handle, CURLINFO_RESPONSE_CODE, &rescode);
-    if (rescode != 200) {
-      LOG_ERROR << "Error pushing root ref, got " << rescode << " HTTP response";
-      errors++;
+  // Push ref
+  if (!is_offline_signed || !push_creds.CanSignOffline()) {
+    if (root_object->is_on_server() == OBJECT_PRESENT) {
+      CURL *easy_handle = curl_easy_init();
+      curl_easy_setopt(easy_handle, CURLOPT_VERBOSE, get_curlopt_verbose());
+      ostree_ref.PushRef(push_server, easy_handle);
+      CURLcode err = curl_easy_perform(easy_handle);
+      if (err) {
+        LOG_ERROR << "Error pushing root ref:" << curl_easy_strerror(err);
+        errors++;
+      }
+      long rescode;
+      curl_easy_getinfo(easy_handle, CURLINFO_RESPONSE_CODE, &rescode);
+      if (rescode != 200) {
+        LOG_ERROR << "Error pushing root ref, got " << rescode << " HTTP response";
+        errors++;
+      }
+      curl_easy_cleanup(easy_handle);
+    } else {
+      LOG_ERROR << "Uploading failed";
+      return false;
     }
-    curl_easy_cleanup(easy_handle);
-  } else {
-    LOG_ERROR << "Uploading failed";
   }
 
   if (errors) {
     LOG_ERROR << "One or more errors while pushing";
-    return EXIT_FAILURE;
+    return false;
   }
 
   if (sign) {
     sign_repo(dst, ostree_ref, hardwareids);
   }
```

Comment:
I'd be happier if sign_repo() was the `else` of the ìf` above that calls `ostree_ref.PushRef`

The error handling here is also done in two different ways. `sign_repo()` uses `std::runtime_error` exceptions to communicate failure, but the rest of the nearby code is using error codes.  I would prefer if we changed sign_repo() to return an error code (and use the `errors` local variable).  (In keeping with "exceptions should be exceptional")

Reply:
Done, but we need sign flag for garage-push so just added check for push_creds.CanSignOffline()